File: app/src/main/java/com/example/jacob/myapplication/Logic/LineAnalyzer.java
package com.example.jacob.myapplication.Logic;

import java.text.SimpleDateFormat;
import java.util.Date;

/**
 * Created by jacob on 20/11/2015.
 */
public class LineAnalyzer {
    SimpleDateFormat sdf = new SimpleDateFormat("dd/MM/yyyy");
    public String getTimeofTheDay(int hour){
        String s = "";
        if (hour >=1 && hour < 6)
            s = "early_morning";
        else if (hour >=6 && hour < 12)
            s = "morning";
        else if (hour >= 12 && hour < 19 )
            s = "afternoon";
        else
            s = "night";

        return s;
    }

    public AnalyzedLine analyzeLine(String line){
        try{

            String[] firstDivision = line.split("-");
            String[] dateDivision = firstDivision[0].split(",");
            String[] textDivision = firstDivision[1].split(":");
            String name = textDivision[0].substring(1,textDivision[0].length());

            String message = textDivision[1];
            Date date =sdf.parse(dateDivision[0]);
            int hour = Integer.parseInt(dateDivision[1].substring(1,dateDivision[1].length()).split(":")[0]);


            return new AnalyzedLine(name, message , date, getTimeofTheDay(hour));

        }
        catch (Exception e){
            System.out.println(e);
            System.out.println("Error: " + line);
            return null;
        }
    }
}

File: app/src/main/java/me/jacobrr/whatsappanalyzer/logic/LineAnalyzer.kt
package me.jacobrr.whatsappanalyzer.logic

import android.util.Log
import java.text.DateFormat.SHORT
import java.text.DateFormat.getTimeInstance
import java.time.Instant
import java.time.LocalDate
import java.time.LocalTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException
import java.time.format.FormatStyle
import java.util.*


/**
 * Created by jacob on 20/11/2015.
 */
class LineAnalyzer {
    private val dateFormatter = DateTimeFormatter.ofLocalizedDate(FormatStyle.SHORT)
    private val timeFormatter = DateTimeFormatter.ofLocalizedTime(FormatStyle.SHORT)
    private val timeFormatter2 = getTimeInstance(SHORT)
    private val regex = Regex("(.+)\\s(\\d{1,2}:\\d{2}.*?)\\s+-\\s+(.+?):\\s+(.+)")
    private var useOldFormatter = false
    private val zone = ZoneId.systemDefault()

    fun getTimeOfTheDay(time: LocalTime): String {
        return when (time.hour) {
            in 1..5 -> "Early morning"
            in 6..11 -> "Morning"
            in 12..18 -> "Afternoon"
            else -> "Night"
        }
    }

    fun analyzeLine(line: String): AnalyzedLine? {
        val lineCleaned = line.replace('\u00A0',' ')
        return try {
            val match = regex.matchEntire(lineCleaned) ?: throw MatchException()

            val name = match.groupValues[3]
            val message = match.groupValues[4]

            val date = parseDate(match.groupValues[1]) ?: throw MyDateTimeException()
            val hour = parseTime(match.groupValues[2]) ?: throw MyDateTimeException()
            AnalyzedLine(name, message, date, getTimeOfTheDay(hour))

        } catch (e: MatchException) {
            Log.e("LINE", "Line does not fit the pattern for a message")
            Log.e("LINE", "Line with error: \"$line\"")
            null
        } catch (e: MyDateTimeException) {
            Log.e("DATE", "Could not parse Date in line $line")
            null
        }
    }

    private fun parseDate(dateStr: String): LocalDate? {
        return try {
            LocalDate.parse(dateStr, dateFormatter)
        } catch (e: DateTimeParseException) {
            Log.e("DATE", "Could not parse Date", e)
            null
        }
    }

    private fun parseTime(dateStr: String): LocalTime? {
        return if (!useOldFormatter) {
            try {
                LocalTime.parse(dateStr, timeFormatter)
            } catch (e: DateTimeParseException) {
                Log.e("TIME", "Could not parse Time with formatter from java.time, format is " +
                        "${LocalTime.now().format(timeFormatter)} and date passed is $dateStr")
                try {
                    useOldFormatter = true
                    Log.d("TIME", "Parsing with old formatter")
                    Instant.ofEpochMilli(timeFormatter2.parse(dateStr)!!.time).atZone(zone).toLocalTime()
                } catch (e: Exception) {
                    Log.e("TIME", "Could not parse time with any formatter, old formatter is " +
                            "${timeFormatter2.format(Date())} and date passed is $dateStr")
                    useOldFormatter = false
                    null
                }
            }
        } else {
            try {
                Instant.ofEpochMilli(timeFormatter2.parse(dateStr)!!.time).atZone(zone).toLocalTime()
            } catch (e: Exception) {
                Log.e("TIME", "Could not parse time with any formatter, old formatter is " +
                        "${timeFormatter2.format(Date())} and date passed is $dateStr")
                null
            }
        }
    }

    class MatchException : Exception("Line does not match regex, is a continuation of the last message")

    class MyDateTimeException : Exception("Could not parse date or time in message")
}
